### PR TITLE
Refactor CoreAttention

### PR DESCRIPTION
**Pull request by hsiehjackson**
* Split long_attention function into different parts and integrate to original attention implementation
  * Modify attention mask/bias
  * Modify query/key/value tokens
  * Apply softmax and get context_layer

### CODE REVIEW

File: nemo/collections/nlp/modules/common/megatron/attention.py [hunk 1]
```diff
@@ -1074,141 +1088,132 @@ def forward(
...
-        # change view [b, np, sq, hn]
-        context_layer = context_layer.view(*output_size)
+            attention_probs = rearrange(attention_probs, 'b np sq sk -> (b np) sq sk')
+            context_layer = torch.bmm(attention_probs, value_layer)
+            context_layer = rearrange(context_layer, '(b np) sq hn -> b np sq hn', np=np)
+            
+        
+        if self.use_long_attention:
+            assert sq == sk, 'Long Attention can only use for self-attention.'
+            nb = (sq // self.local_context)
```
Comment (sam1373) [inline nemo/collections/nlp/modules/common/megatron/attention.py]: need to subtract total_transient tokens or nb may be wrong

File: nemo/collections/nlp/modules/common/megatron/attention.py [hunk 2]
```diff
@@ -1074,141 +1088,132 @@ def forward(
...
-
-            # Raw attention scores. [b, np * sq, sk]
-            matmul_result = torch.baddbmm(
-                matmul_input_buffer,
-                query_layer,  # [b, np * sq, hn]
-                key_layer,  # [b, hn, sk]
-                beta=0.0,
-                alpha=(1.0 / self.norm_factor),
+            query_layer = rearrange(query_layer, 'sq b np hn -> b (np sq) hn')
+            key_layer = rearrange(a, 'sk b 1 hn -> b hn sk')
```
Comment (sam1373) [inline nemo/collections/nlp/modules/common/megatron/attention.py]: what is a? typo?

File: nemo/collections/nlp/modules/common/megatron/attention.py [hunk 3]
```diff
@@ -1031,40 +1021,64 @@ def forward(
...
+                    :,
+                    self.num_attention_heads_partition_offset : self.num_attention_heads_partition_offset
+                    + self.num_attention_heads_per_partition,
+                ]
+                if attention_bias.shape[0] == 1:
+                    attention_bias = attention_bias.expand(b, *([-1]*(attention_bias.dim() - 1)))
+            else:
+                # relative_position_bias is T5RelativePositionEmbedding
+                attention_bias = self.get_local_pos_bias(
+                    relative_position_bias, nb, self.local_context
```
Comment (sam1373) [inline nemo/collections/nlp/modules/common/megatron/attention.py]: nb isn't defined yet so this fails. though we are currently never using dynamic pos bias so maybe could remove this whole section for now

File: nemo/collections/nlp/modules/common/megatron/attention.py [hunk 4]
```diff
@@ -1313,189 +1317,91 @@ def long_attention(
...
+        ).expand(-1, np, -1, -1, -1)
 
-        attention_mask = torch.empty(
-            batch_size,
-            nh,
-            nb,
-            sq,
-            sk,
+        # Default attend to all tokens including global tokens
+        attention_mask = torch.zeros(
```
Comment (sam1373) [inline nemo/collections/nlp/modules/common/megatron/attention.py]: I was wrong before, this should be ones as we're using True to mean attend here

**Review by sam1373 — COMMENTED** [review]
overall great job

**Comment by sam1373** [comment]
also pls resolve conflict

File: nemo/collections/nlp/modules/common/megatron/attention.py (hunk 1 above)
Comment (hsiehjackson) [reply nemo/collections/nlp/modules/common/megatron/attention.py]: We already add the transient tokens to `query_layer` and `key_layer` before pass in `CoreAttention`?

Comment (sam1373) [reply nemo/collections/nlp/modules/common/megatron/attention.py]: yes, because we average hidden for creating transient kv, and we don't pass hidden to here
